pkg/k8sapi/discovery.go

```diff
@@ -68,15 +68,31 @@ type Controller struct {
 	controller *controller
 	kubeClient kubernetes.Interface
 	myNodeName string
-	synced     bool
 }
 
 // NewController creates a new DiscoveryController
-func NewController(clientset kubernetes.Interface) *Controller {
-	return &Controller{kubeClient: clientset,
+func NewController(clientset kubernetes.Interface) (*Controller, error) {
+	stopCh := wait.NeverStop
+
+	d := &Controller{kubeClient: clientset,
 		myNodeName: os.Getenv("MY_NODE_NAME"),
 		cniPods:    make(map[string]string),
 		workerPods: make(map[string]*K8SPodInfo)}
+
+	log.Info("Starting Pod informer")
+
+	go d.controller.informer.Run(stopCh)
```

Comment:
`d.controller` will be `nil` here, since `d.controller` is created in the `DiscoverK8sPods()` method:

https://github.com/aws/amazon-vpc-cni-k8s/blob/7c6dec755c61703b966799cc5af26012044a9410/pkg/k8sapi/discovery.go#L155

What you need to do is basically get rid of the `DiscoverK8SPods()` method entirely, and move the code that is in there into the `NewController()` function.

The only reason that the `DiscoverK8SPods()` method even exists as a separate struct method apparently is to "start the controller" by doing:

```go
select {}
```

but that's really kinda hacky IMHO. A cleaner solution would be to simply move all this setup code into `NewController()`:

```go
	// create the pod watcher
	podListWatcher := cache.NewListWatchFromClient(d.kubeClient.CoreV1().RESTClient(), "pods", metav1.NamespaceAll, fields.OneTermEqualSelector("spec.nodeName", d.myNodeName))

	// create the workqueue
	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())

	// Bind the workqueue to a cache with the help of an informer. This way we make sure that
	// whenever the cache is updated, the pod key is added to the workqueue.
	// Note that when we finally process the item from the workqueue, we might see a newer version
	// of the Pod than the version which was responsible for triggering the update.
	indexer, informer := cache.NewIndexerInformer(podListWatcher, &v1.Pod{}, 0, cache.ResourceEventHandlerFuncs{
		AddFunc: func(obj interface{}) {
			key, err := cache.MetaNamespaceKeyFunc(obj)
			if err == nil {
				queue.Add(key)
			}
		},
		UpdateFunc: func(old interface{}, new interface{}) {
			key, err := cache.MetaNamespaceKeyFunc(new)
			if err == nil {
				queue.Add(key)
			}
		},
		DeleteFunc: func(obj interface{}) {
			// IndexerInformer uses a delta queue, therefore for deletes we have to use this
			// key function.
			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
			if err == nil {
				queue.Add(key)
			}
		},
	}, cache.Indexers{})

	d.controller = newController(queue, indexer, informer)
```

and move this code:

```go
	// Now let's start the controller
	stop := make(chan struct{})
	defer close(stop)
	go d.run(1, stop)

	// Wait forever
	select {}
```

to a separate `Controller.Listen()` method that you can call after returning successfully from `NewController()`